utils/data_loader.py

```diff
@@ -423,3 +425,90 @@ def dataset_1(self):
         return self._dataset_2
+
+
+class ContextualizedRankingDataset(Dataset):
+
+    def __init__(self, data_path, bert_model, max_len, lower_case, batch_size, separator, mod, head,
+                 label, label_definition_path):
+        """
+        This Dataset can be used to train a composition model with contextualized embeddings to create attribute-like
+        representations
+        :param data_path: [String] The path to the csv datafile that needs to be transformed into a dataset.
+        :param bert_model: [String] The Bert model to be used for extracting the contextualized embeddings
+        :param max_len: [int] the maximum length of word pieces (can be a large number)
+        :param lower_case: [boolean] Whether the tokenizer should lower case words or not
+        :param separator: [String] the csv separator
+        :param label: [String] the label of the column the class label is stored in
+        :param mod: [String] the label of the column the modifier is stored in
+        :param head: [String] the label of the column the head is stored in
+        :param label_definition_path: [String] path to the file that holds the definitions for the labels
+        """
+        self._data = pandas.read_csv(data_path, delimiter=separator, index_col=False)
+        self._definitions = pandas.read_csv(label_definition_path, delimiter="\t", index_col=False)
+        self._modifier_words = list(self.data[mod])
+        self._head_words = list(self.data[head])
+        self._phrases = [self.modifier_words[i] + " " + self.head_words[i] for i in range(len(self.data))]
+        self._labels = list(self.data[label])
+        self._label2definition = dict(zip(list(self._definitions["label"]), list(self._definitions["definition"])))
+        self._label_definitions = [self._label2definition[l] for l in self.labels]
+        assert len(self.modifier_words) == len(self.head_words) == len(
+            self.phrases), "invalid input data, different lenghts"
+
+        self._feature_extractor = BertExtractor(bert_model=bert_model, max_len=max_len, lower_case=lower_case,
+                                                batch_size=batch_size)
+        self._samples = self.populate_samples()
+
+    def lookup_embedding(self, simple_phrases, target_words):
+        return self.feature_extractor.get_single_word_representations(target_words=target_words,
+                                                                      sentences=simple_phrases)
+
+    def populate_samples(self):
+        """
+        Looks up the embeddings for all modifier, heads and labels and stores them in a dictionary
+        :return: List of dictionary objects, each storing the modifier, head and phrase embeddings (w1, w2, l)
+        """
+        word1_embeddings = self.lookup_embedding(target_words=self.modifier_words, simple_phrases=self.phrases)
+        word2_embeddings = self.lookup_embedding(target_words=self.head_words, simple_phrases=self.phrases)
```

Comment:
is it not better to include here the possibility to choose a context sentence instead of the phrase? I assume we will do this later and it will have to be changed again then.

Reply:
done..